Track multi-confidence filtering/displaying

BryonLewis opened the pull request:
Fixes #423 

Old style didn't properly support multiple types on a single track and swapping between them based on the selected types and the confidence filter adjustments.

- Added a local variable `currentConfidencePair` to indicate the index of the sorted confidence pairs that should be displayed.
- Within `useTrackFilters` I modified the filtering function to set the `currentConfidencePair` when it is visible and above the `confidenceThreshold`. Let me know how you feel about having a side effect inside of `.some`.
- Returned the confidencePairs[currentConfidencePair] when using `track.getType()` so that it will render the proper type when displayed in `AnnotationLayers`, and `trackList`.
- Updated `useEventChart` and `useLineChart` to utilize the `track.getType()` function instead of the `confidencePairs[0][0]` so they would reflect the proper type as well.
- Added some tests to the `track.spec.ts` to test setting the currentConfidencePair, as well as changing the current type.

Old Version:
![oldConfidenceSort](https://user-images.githubusercontent.com/61746913/100242173-7d168500-2f02-11eb-9d79-e000f8397aea.gif)
New Version:
![newConfidenceSort](https://user-images.githubusercontent.com/61746913/100242185-8273cf80-2f02-11eb-868f-3c70ecce155c.gif)

Questions : 
(made a determination that these questions will be placed in a new PR or as new issues, I just want to fix the fundamental problem first in a smaller PR.  Feel free to add comments)

- Determine what we want to do when deleting tracks by type?  Do we delete the type from the track, or do we delete the whole track?
- Should we display all types as text in the upper right corner?  In the older version of VIAME-Web all types were displayed as a list in the upper right corner of the annotation.  This can be done with some minor modifications.

subdavis commented:
> Determine what we want to do when deleting tracks by type? Do we delete the type from the track, or do we delete the whole track?

Delete type until only 1 remains, then delete track (IMO).

> Should we display all types as text in the upper right corner? In the older version of VIAME-Web all types were displayed as a list in the upper right corner of the annotation. This can be done with some minor modifications.

Yeah I think we should.  It would be nice to stack them instead of side-by-side, I think.

subdavis reviewed (changes requested):
I had to go pause and have a plate of cold thanksgiving while I thought about this.

## the problem

The problem is communicating a new bit of state to consumers of the `filteredTracks` computed prop.  Places that use this prop need to know something new: "What is the confidence pair in this track that caused it be included in this list".  Before, that was implicit.  There was only 1 possibility.  Now there are several.

You've implemented this by sort of stashing this information inside the track object.

I object to that for a couple reasons:

* It improperly decouples `filteredTracks` from the `currentConfidencePair`.  You don't care about `currentConfidnecePair` as context unless you already have `filteredTracks`.   
* It adds new state/methods to the track API that aren't really related to a track itself.
* Side effects.

## Alternative.

It seems to me like the best outcome would be to change the shape of `filteredTracks` to include this new bit of information. 

``` typescript
const filteredTracks: Ref<{
  track: Track,
  context: {
    confidencePairIndex: number;
  };
}[]>
```

This involves either changing the interface in a bunch of places or having another computed property that strips out just the track part for interfaces where the context isn't needed.  IMO it would be best to just change the interface everywhere.

Naturally the enabledTracks interface will change also.  I think that's fine.  This new context is nice because any process that needs to explain why it filtered something can just tack on new state, and modules that accept a filtered list can specify which explanations they need.

In the short term, if you want to just declare this `interface FilterItem {}` at the top of `useTrackFilters` and import it from everywhere, that's perfectly fine with me.

If you would like me to do the tedious chore of changing the interface everywhere, I am happy to do so.

 ----

You could also "re-derive" this information in-place whereever you need it.  If you have a track, and you have the list of enabledTypes, you can easily figure out which of a track's types caused its inclusion.  However, this does involve subtly duplicating some logic which could lead to bugs.  I don't recommend this option.

subdavis commented:
Thought about it some more today.

If you want to do something easier like a dedicated track context, what about this?

```
class Track {
  /* Disclaimer about usage and reactivity */
  context: Map<string, string | number>;
}
```

I don't think methods to get/set the map are necessary.  just `track.context.get('key')` and `track.context.set('key', 'value')`

The map sort of guarantees it's not reactive, and the lack of methods guarantees we aren't bumping the notifier.

BryonLewis commented:
Getting back into this I want to attempt to accurately summarize this and the possible implications.

Summarize the highlights:

- useTracks has been changed to useFilteredTracks and has a structure indicating the `confidencePairIndex` which is the current confidence pair that should be visible based on the confidence filter and the visibility of the types.
- To accommodate this the `track.getType()` function now takes in an argument being the index.  This allows the current ConfidencePair index to be shown based on the useFilteredTracks.
- In LayerManager the dataTrack item has been modified to provide the `trackType` which is the current type of the track provided by useFilteredTracks, and confidencePairs has been changed to represent the list of `confidencePairs`, I think it did that originally but at some point it was swapped to just the first confidencePair in the list.  I think these variable names make more sense.  You have the full list of confidencePairs and the currentPair provided by trackType.
- The change to FrameDataTrack required updating the rendering code in each of the layers. This is where it could impact subscribers of vue-media-annotator depending on how they are using this
- The TextLayer has been modified to draw all of the confidence pairs for a track in order with the current item being brighter and the other ones have a lower opacity.
- There are now helper functions when deleting visible types which will remove confidence pairs from the track.  This will also be used when we implement the confidence pair stuff in the `TrackDetailsEditor` (The old Attributes view for a track)

subdavis commented on client/src/layers/TextLayer.ts:
```diff
@@ -11,32 +11,70 @@ export interface TextData {
  * @returns value or null.  null indicates that the text should not be displayed.
  */
-function defaultFormatter(track: FrameDataTrack): TextData[] | null {
+function defaultFormatter(track: FrameDataTrack, additionalNum = 0): TextData[] | null {
```
`additionalNum` is sort of an awkward way of specifying a max value.  Does it mean 1 will always be shown, plus some `additionalNum`? 

Could we just make this `maxLines` or `maxPairs` and default it to 1?  I think that's clearer, then you wouldn't have to do `(maxPairs + 1)` below (which is odd anyway because it means that maxPairs = 0 will show 1 line)

BryonLewis replied:
I think restructuring this will make it clearer by moving and renaming `maxPairs` to `maxAdditionalPairs` down closer to where it is drawn instead of with the default type:confidence pair.

subdavis commented on client/src/layers/TextLayer.ts:
```diff
@@ -11,32 +11,70 @@ export interface TextData {
+      //Now we display any additional types besides the default type
+      currentHeight += lineHeight;
+      let pairCount = 0;
+      if (track.confidencePairs.length) {
```
No need for this.  For loop over `[]` won't run.

subdavis commented on client/src/layers/TextLayer.ts:
```diff
@@ -11,32 +11,70 @@ export interface TextData {
+      const maxPairs = Math.min(track.confidencePairs.length, additionalNum);
+      let currentHeight = bounds[1] - (lineHeight * (maxPairs + 1));
+      const arr: TextData[] = [];
+      const baseType = track.trackType ? track.trackType[0] : 'unknown';
+      const baseConfidence = track.trackType ? track.trackType[1] : 1.0;
```
I don't think this is a good place to do defaults.

Tracks without confidence pairs shouldn't be allowed at all into this application.  `unknown: 1.0` defaults should be forced at ingest or, really, by the backend.  I actually think the frontend should refuse to load at all and show an error message, because a track without confidence pairs is malformed data.

This seems out-of-scope for this PR.

subdavis commented on client/src/use/useLineChart.ts:
```diff
@@ -42,14 +42,15 @@ export default function useLineChart({
-      if (confidencePairs.length) {
-        const trackType = confidencePairs[0][0];
+      const type = track.getType(filtered.context.confidencePairIndex);
+      if (type && type.length) {
```
Overly defensive I think, and contradicts the method return type.

```
getType(index = 0): [string, number] | null {
```

`if (type)` is enough, because if it's not null, it must be a `[string, number]`.

Pending discussion, even that check might be unnecessary.

subdavis commented on client/src/use/useLineChart.ts:
```diff
@@ -42,14 +42,15 @@ export default function useLineChart({
-        const trackType = confidencePairs[0][0];
+      const type = track.getType(filtered.context.confidencePairIndex);
+      if (type && type.length) {
+        const trackType = type ? type[0] : '';
```
```suggestion
        const trackType = type[0];
```

also too defensive.  This should throw an out-of-bounds if it's wrong, not fail silently.  Similar to above note about not having defaults for invalid states.

subdavis commented on client/src/use/useTrackFilters.ts:
```diff
@@ -61,24 +67,27 @@ export default function useFilteredTracks(
-        /* include tracks with no confidence pairs */
-        || track.confidencePairs.length === 0
-      );
+      if (confidencePairIndex !== -1 || track.confidencePairs.length === 0) {
```
```suggestion
      if (confidencePairIndex >= 0 || track.confidencePairs.length === 0) {
```

Nit.  Include vs exclude condition.

subdavis commented on client/src/layers/TextLayer.ts:
```diff
@@ -11,32 +11,70 @@ export interface TextData {
+      let pairCount = 0;
+      if (track.confidencePairs.length) {
+        for (let i = 0; i < track.confidencePairs.length; i += 1) {
+          if (pairCount >= maxPairs) {
```
If you change `maxPairs`, this line has to change, but I think it should because this isn't what you'd expect if `maxPairs` wasn't already off-by-one.

subdavis commented on client/src/track.ts:
```diff
@@ -341,13 +341,28 @@ export default class Track {
-  getType(): [string, number] | null {
+  getType(index = 0): [string, number] | null {
     if (this.confidencePairs.length > 0) {
+      if (this.confidencePairs[index]) {
+        return this.confidencePairs[index];
+      }
```
This doesn't seem right.

if confidencePairs is `[['foo', 0.5], ['bar', 0.6]]` and I call getType(20000), I'll get `['foo', 0.5]`.  I think I should get `null`.

BryonLewis replied:
I was maintaining parity with the existing getType() which returned the 0 indexed confidencePair if it existed or else it would return `null`.  So I modified it to return the 0 index confidencePair by default if it existed and `index` was out of scope.  Or else it would return null if the confidencePairs were empty.

We can change this but I'll need to review and understand why we had it like this in the first place.  I imagine it has something to do with creation or CSV files which don't always contain types.

subdavis commented on client/src/use/useEventChart.ts:
```diff
@@ -26,10 +27,12 @@ export default function useEventChart({
       if (confidencePairs.length) {
-        const trackType = confidencePairs[0][0];
+        const type = track.getType(filtered.context.confidencePairIndex);
+        const trackType = type ? type[0] : '';
```
This is causing me to question the signature of `track.getType()`

It should probably always return `[string, number]` instead of the possibility of a null type, right?  Like, if you ask for an out-of-bounds pair, that seems like an invalid execution path that should also throw an error from within the getter.

Why would `getType` ever be null?  Why would you ever have a reference to a confidencePair index that doesn't exist?

subdavis reviewed (changes requested):
I can't figure out how to get the viewer to show more than 1 confidence pair.  Is there a setting somewher?  Enabling / disabling types works as expected (the next highest type appears), but I only ever see one even though the code for multiple is clearly here.

subdavis commented on client/src/components/LayerManager.vue:
```diff
@@ -43,7 +44,7 @@ export default defineComponent({
     const intervalTree = useIntervalTree();
     const trackMap = useTrackMap();
-    const tracksRef = useEnabledTracks();
+    const filteredTracksRef = useEnabledTracks();
```
I know this will need to change in a lot of places, but having `filteredTracks` mean several different things  is probably not a good idea.

BryonLewis commented:
> I can't figure out how to get the viewer to show more than 1 confidence pair. Is there a setting somewher? Enabling / disabling types works as expected (the next highest type appears), but I only ever see one even though the code for multiple is clearly here.

![image](https://user-images.githubusercontent.com/61746913/108735100-2408be00-74fe-11eb-8821-c7e3c8814074.png)
Here I set `additionalNum` to 2.  The current value is closest to the box and more opaque than the additional values added.  It also has the ** next to it which can be removed or changed.

It should be the default value of `additionalNum` in the `defaultFormatter` function.  By default we always want to display the highest confidence pair for the text value.  I don't know the exact format we want for additional values to be displayed.  Do we want the top X values?  If it has 200 values and we are filtered down to item number 50, do the top 5 values show any relevance?  Should we take the closes X values in distance to the current value.  Meaning do we grab x/2 above and below the current value and only below if this is the top value already?  I left the display code in there as a beginning for supporting more visually.  I y believe this needs to be decided before we activate it as well as allowing users to specify themselves how many additional pairs they see.

BryonLewis commented:
@subdavis hopefully I didn't botched this up too horribly.  I believe that I addressed all of your comments and tried to make things clearer.  After looking at what calls `getType` there seems to be no reason for it to not through an error if the `confidencePairs[index]` doesn't exist.  We will just need to ensure in the future on import that all tracks have a default pair (unknown, 1.0) if nothing exists.

subdavis commented:
Pending your thoughts on #603, this is good to go.

subdavis reviewed (approved):
Thanks for making these changes.  I think just changing the names for `enabledTracks` and `TrackWithContext` are the biggest improvements.
